design/dedicated-osd-pod.md

```diff
@@ -0,0 +1,66 @@
+- Process management issue. Rook's process management is very simple. Using Kubernetes pod management is much more reliable.
+
+
+A more comprehensive discussions can be found at [this issue](https://github.com/rook/rook/issues/1341).
+
+## Proposal   
+
+We propose the following change to address the limitations.
+
+### Terms
+
+- OSD Discovery. A process that is given a device or directory path filter and use the filter to discover devices on the host. The process updates the OSD orchestration Configmap with the device paths and mark the orchestration status as "DeviceDiscovered". The process is created by Rook Operator and runs in a Kubernetes deployment on the storage node. This process is storage backend agnostic. Currently `rook osd` command implements this functionality.
+
+- OSD Provisioner. A process that is given device or directory paths upon start and make backend specific storage types. For now, the process prepare Ceph OSDs. Currently `rook osd` command implements this functionality.
+
+### Create new OSDs
+| Sequence |Rook Operator | OSD Discovery  | OSD Provisioner   | Ceph OSD Deployment | 
+|---|---|---|---|---|
+| 1  | Read devices from cluster CRD and create an OSD Discovery deployment on each storage node  |   |   |
+| 2  | Watch Configmap  | Discover OSD devices that fit those defined in Cluster CRD and update Configmap with the device paths| | |
+| 3  | Detect Configmap change, parse Configmap, extract device paths, and create an OSD Provisioner deployment for each device  |Detect device additions or removals | | |
+| 4  | Watch Configmap |Detect device additions or removals | Prepare OSDs, Persist OSD ID, datapath, and node info in a per node Configmap | |
+| 5  | Detect Configmap change, parse Configmap, extract OSD info, construct OSD Pod command and arg | Detect device additions or removals | | |
+| 6  | Create one deployment per OSD |Detect device additions or removals | |
+| 7  | |Detect device additions or removals | | Start `ceph-osd` Daemon one Pod per device |
+
+This change addresses the above limitations in the follow ways:
+- High reliability. Each `ceph-osd` daemon runs its own Pod, their restart and upgrade are by Kubernetes controllers. Upgrading OSD Provisioner Pod no longer restarts `ceph-osd` daemons.
+- More efficient resource requests. Once OSD Discovery detects all OSD devices, Rook Operator is informed of the topology and assigns appropriate resources to each Ceph OSD deployment.
+
+### Discussions
+
+It is expected OSD Discovery process will be merged into Rook Operator once local PVs are supported in Rook Cluster CRD. Rook Operator can infer the device topoloyg from local PV Configmaps. 
+
+Alternatively, since `rook agent` is currently running as a DaemonSet on all nodes, it is conceivable to make `rook agent` to poll devices and update OSD orchestration Configmap. This approach, however, needs to give `rook agent` the privilege to modify Configmaps.
```

Comment:
I think i'm a fan of just using the rook-agent to perform the discovery.  It is already being deployed to every machine in the cluster and it already has access to the raw devices of nodes in the cluster.  And it's a long running daemon.  It feels like a great candidate for performing the discovery work as well instead of adding another component to do it instead.

Reply:
@jbw976 thanks for the comment. `rook agent` is indeed a convenient candidate to discover devices. It already bind mounts hosts `/dev` and long running. 

Device Discovery Pod should have minimal privileges possible, while `rook` agent is designed to accommodate many deployment requirements where privileges are required. 

I am concerned of the following security models:
- `rook agent` is running in [privileged mode](https://github.com/rook/rook/blob/master/pkg/operator/agent/agent.go#L98). This is not needed for `Device Discovery Pod`. 
- `rook agent` needs to bind mount hosts `/dev` with read/write mode, while `Device Discovery Pod` only needs read only.
-  `Device Discovery Pod` needs to get and update Configmaps, while `rook agent` only needs to [get but not update Configmaps](https://github.com/rook/rook/blob/master/pkg/operator/agent/agent.go#L52).

Morever,  `rook agent` [may not run on storage node](https://github.com/rook/rook/blob/master/pkg/operator/agent/agent.go#L190). For topology where compute and storage are separated, `rook agent` only runs on compute nodes but not storage nodes. In such case, assumption that `rook agent` runs on storage nodes doesn't hold.